### PR TITLE
Server Logs Read Fix

### PR DESCRIPTION
**Pull request by valaises**

### CODE REVIEW

**Comment by olegklimov** [comment]
live streaming -n 1000 is a good idea

for plain file 50000 maybe is a good idea too

File: refact_webgui/webgui/tab_server_logs.py [hunk 1]
```diff
@@ -1,13 +1,48 @@
...
 
+import aiofiles
 from fastapi import APIRouter
 from fastapi.responses import StreamingResponse, Response
 from refact_utils.scripts import env
 
 
+async def count_lines_wc(file_path: str) -> int:
+    cmd = ["wc", "-l", file_path]
+    print(f"EXEC: {' '.join(cmd)}")
```
Comment (mitya52) [inline refact_webgui/webgui/tab_server_logs.py:15]: why print? we have logger if it really heeds the logging

File: refact_webgui/webgui/tab_server_logs.py [hunk 2]
```diff
@@ -1,13 +1,48 @@
...
+    stdout, stderr = await process.communicate()
+
+    if process.returncode == 0:
+        return int(stdout.decode().split()[0])
+
+    raise Exception(f"{' '.join(cmd)} command failed: {stderr.decode()}")
+
+
+async def tail(file_path: str, last_n_lines: int) -> AsyncIterator[str]:
+    cmd = ["tail", f"-n {last_n_lines}", file_path]
```
Comment (mitya52) [inline refact_webgui/webgui/tab_server_logs.py:30]: is there a python package for this task?

File: refact_webgui/webgui/tab_server_logs.py (hunk 1 above)
Comment (mitya52) [reply refact_webgui/webgui/tab_server_logs.py:15]: and is there a package for count lines in file?

File: refact_webgui/webgui/tab_server_logs.py [hunk 3]
```diff
@@ -48,7 +74,28 @@ async def get_log(right_file):
             return Response("File \"%s\" not found\n" % log_name, media_type="text/plain")
         right_file = list_of_files[-1]
 
+        async def streamer() -> AsyncIterator[str]:
+            last_n_lines = 10_000
```
Comment (mitya52) [inline refact_webgui/webgui/tab_server_logs.py:78]: make it default parameter instead of var

Comment (mitya52) [inline refact_webgui/webgui/tab_server_logs.py:78]: make it default parameter instead of var

File: refact_webgui/webgui/tab_server_logs.py [hunk 4]
```diff
@@ -48,7 +74,28 @@ async def get_log(right_file):
...
+
+            if (lines_cnt := await count_lines_wc(right_file)) > last_n_lines:
+                yield f'{lines_cnt - last_n_lines} lines were skipped...\n'
+
+            async with aiofiles.open(right_file, 'r') as f:
+                # going to the end of the file before streaming lines_from_tail -- while we are streaming some new lines might appear
+                await f.seek(0, os.SEEK_END)
```
Comment (mitya52) [inline refact_webgui/webgui/tab_server_logs.py:85]: what is the purpose of this line? why we should go to the end of the file if we simply need to stream last n lines?

File: refact_webgui/webgui/tab_server_logs.py [hunk 5]
```diff
@@ -48,7 +74,28 @@ async def get_log(right_file):
...
+                async for line in tail(right_file, last_n_lines):
+                    yield line
+
+                if stream:
+                    while True:
+                        line = await f.readline()
+                        if not line:
+                            await asyncio.sleep(0.1)
+                            continue
+                        yield line
```
Comment (mitya52) [inline refact_webgui/webgui/tab_server_logs.py:96]: this looks weird: you're streaming last lines with tail (on opened file!!!) and then read from file directly...

**Review by mitya52 — CHANGES_REQUESTED** [review]

File: refact_webgui/webgui/tab_server_logs.py [hunk 6]
```diff
@@ -48,7 +43,16 @@ async def get_log(right_file):
             return Response("File \"%s\" not found\n" % log_name, media_type="text/plain")
         right_file = list_of_files[-1]
 
+        async def streamer(last_n_lines: int = 10_000) -> AsyncIterator[str]:
+            for line in tailer.tail(open(right_file, 'r'), last_n_lines):
+                yield line + '\n'
+
+            if stream:
+                async for line in atail(right_file):  # (line, from_file)
```
Comment (mitya52) [inline refact_webgui/webgui/tab_server_logs.py:51]: here we can loose some lines from working logs

Comment (mitya52) [reply refact_webgui/webgui/tab_server_logs.py:51]: inthe case of streaming it should work exactly as
tail -n last_n_lines -f logfile

**Review by mitya52 — CHANGES_REQUESTED** [review]

**Comment by mitya52** [comment]
doesn't work correctly with
* non-existed file or non-file paths: stuck without exit
* stream=False doesn't work for regular log files, while loops forever

**Pull request closed without merging**